### PR TITLE
Add serial support to the PET

### PR DESCRIPTION
**Pull request by rbergen**
This adds a serial driver for the PET, thereby enabling visualization of real audio data on the PET. It uses the relevant parts of [PETTERM](https://github.com/ChartreuseK/PETTERM) as a basis for this.

Note that a few new settings have been added to settings.inc. One relates to the transmission of the * back to the PET, to indicate that a redraw of VU meters and bars has occurred. On PETs with a built-in piezo speaker, that transmission causes the speaker to make a clicking noise. So, on those PETs one can choose to either report back framerates or enjoy silent PET Rock operation.

Also part of this PR:

- An overall restructuring of the code, largely to facilitate the presence of serial drivers for both C64 and PET.
- An optimization of the drawing logic in the sense that VU meters and bars are only actually drawn when there is (new) data to draw.

### CODE REVIEW

File: serial/pet/vars.s [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+KbdPollIntrvl:  .res 1      ; KBD Polling interval for baud
+KbdPollCnt:     .res 1      ; Polling interval counter
+KbdByte:        .res 1      ; (Last) byte read from keyboard
+KeyCol:         .res 1      ; Key scan column
+KeyRow:         .res 1      ; Key row that is being scanned
+KeyRowFound:    .res 1      ; Key row in which pressed key was found
+KeyBitsFound:   .res 1      ; Bits of key that was found depressed
+KeyOffset:      .res 1      ; Keyboard matrix offset for shift
+KbdTemp:        .res 1      ; Keyboard scanning temp, to allow BIT instruction
+KbdNewFlag:	    .res 1      ; Indicate if new key was pressed
```
Comment (davepl) [inline serial/pet/vars.s:21]: Minor formatting issue

File: serial/pet/driver.s [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+                sta VIA_CR
+
+                lda #$EC              ; Tx as output high, uppercase+graphics ($EE for lower)
+                                      ; CA1 trigger on falling edge
+                sta VIA_PCR
+                ; Set VIA interrupts so that our timer is the only interrupt source
+                lda #$7F              ; Clear all interrupt flags
+                sta VIA_IER
+                lda #$C2              ; Enable Timer 1 interrupt and CA1 interrupt
+                sta VIA_IER
```
Comment (davepl) [inline serial/pet/driver.s:91]: Just want to make sure this is what is intended.  I assume it latches it or something but its two stores to the same location./

File: serial/pet/driver.s [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+                sta RxBufReadPtr
+
+                rts
+
+;-----------------------------------------------------------------------------------
+; CloseSerial: Teardown serial comms. We try to restore everything to its original
+; state.
+;-----------------------------------------------------------------------------------
+
+CloseSerial:
```
Comment (davepl) [inline serial/pet/driver.s:160]: A few comments on what all these masks and values are doing as they are stored in the PIA/VIA might be prudent?

File: serial/pet/driver.s [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+                ldx RxBufWritePtr
+                lda RxCurByte
+                sta RX_BUF,X
+
+                inc RxBufWritePtr
+
+                lda #$22              ; Disable timer 2 interrupt and CA1
+                sta VIA_IER
+                lda #$82              ; Enable CA1 interrupt
+                sta VIA_IER
```
Comment (davepl) [inline serial/pet/driver.s:533]: Same here again, just confirming that intention is to his the same register twice

File: serial/pet/driver.s [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+                ; There are backward of the table above! Above goes from 9->0, these are 0->9
+KeyMask:        .byte $FF, $BF, $FF, $FF, $FF, $FF, $BE, $FF, $FE, $BF
+                ; Which bits indicate shift keys
+ShiftMask:      .byte $00, $00, $00, $00, $00, $00, $41, $00, $00, $00
+                ; Which bits indicate ctrl keys
+CtrlMask:       .byte $00, $00, $00, $00, $00, $00, $00, $00, $01, $00
+
+                ; Keyboard matrix with shift pressed, needed for consistent shifts
+                ; Matrix for Business Keyboards
+KbdMatrixShift: .byte '>', $0E, $F4, '8', '=', '(', '%', '"'    ;" ;$8E = BothShift+2, $9D = CursRight
```
Comment (davepl) [inline serial/pet/driver.s:682]: I'm just going to assume this is right :-)

**Review by davepl — APPROVED** [review]
A few minor comments inline but nothing wrong that I can spot!

File: serial/pet/vars.s (hunk 1 above)
Comment (rbergen) [reply serial/pet/vars.s:21]: I don't think I understand what minor formatting issue you're referring to? As far as I can tell, spacing and spelling are consistent.

File: serial/pet/driver.s (hunk 2 above)
Comment (rbergen) [reply serial/pet/driver.s:91]: The first write explicitly disables all interrupts (bit 7 in the value written is 0), the second then explicitly enables two interrupts (bit 7 written is 1). The only way to make sure you achieve the desired result of having only two interrupts enabled is two consecutive writes.

File: serial/pet/driver.s (hunk 3 above)
Comment (rbergen) [reply serial/pet/driver.s:160]: I've copied this code verbatim from the PETTERM tear down code, so I could pass that comment on to the authors of PETTERM. :) I assume these values are all retrieved from the PET KERNAL initialization code, either by copying parts of the KERNAL ROM or establishing what the respective values are immediately after the machine starts up.

File: serial/pet/driver.s (hunk 4 above)
Comment (rbergen) [reply serial/pet/driver.s:533]: Yes, the first write disables two interrupts (bit 7 is 0), the second enables one (bit 7 is 1).

File: serial/pet/driver.s (hunk 5 above)
Comment (rbergen) [reply serial/pet/driver.s:682]: As did I. :) We can't really test this without a PET with a business keyboard, I think.

**Pull request merged**